File: backend/app/schemas/monitoring.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.domain.enums import (
    AdvisorTriggerStatus,
    BehaviorInterventionStatus,
    BehaviorObservationType,
    MonitoringAlertStatus,
    MonitoringCadence,
    MonitoringComparator,
    MonitoringPolicyType,
    MonitoringSeverity,
    RiskLimitEffect,
)


class BehaviorObservationInput(BaseModel):
    model_config = ConfigDict(extra="forbid")

    observation_type: BehaviorObservationType
    signal_strength: Decimal = Field(ge=0, le=1, max_digits=9, decimal_places=6)
    occurrence_count: int = Field(default=1, ge=1, le=10000)
    observed_at: datetime | None = None
    source_reference: str = Field(min_length=1, max_length=160)
    financial_event_id: str | None = None
    evidence: dict[str, Any] = Field(default_factory=dict)


class MonitoringEvaluateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    analysis_date: date | None = None
    market_shock: bool = False
    hard_facts_changed: bool = False
    customer_confirmation_id: str | None = None
    observations: list[BehaviorObservationInput] = Field(default_factory=list, max_length=100)
    is_user_confirmed: Literal[True]


class MonitoringPolicyOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    policy_type: MonitoringPolicyType
    metric: str
    comparator: MonitoringComparator
    threshold: Decimal
    cadence: MonitoringCadence
    severity: MonitoringSeverity
    active: bool
    effective_from: date
    effective_to: date | None
    rule_version: str


class MonitoringAlertOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    household_id: str
    monitoring_policy_id: str
    financial_event_id: str | None
    policy_type: MonitoringPolicyType
    trigger_reason: str
    client_impact: str
    severity: MonitoringSeverity
    recommended_action: str
    do_not_sell_flag: bool
    required_specialist: str | None
    evidence_snapshot: dict[str, Any]
    status: MonitoringAlertStatus
    detected_at: datetime
    resolved_at: datetime | None


class MonitoringAlertsResponse(BaseModel):
    household_id: str
    rule_version: str
    alerts: list[MonitoringAlertOut]


class BehaviorObservationOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: str
    observation_type: BehaviorObservationType
    observed_at: datetime
    signal_strength: Decimal
    occurrence_count: int
    risk_limit_effect: RiskLimitEffect
    hard_facts_changed: bool
    source_reference: str
    evidence_snapshot: dict[str, Any]


class MonitoringEvaluateResponse(BaseModel):
    household_id: str
    analysis_date: date
    rule_version: str
    evaluated_policy_count: int
    triggered_policy_count: int
    resolved_alert_count: int
    policies: list[MonitoringPolicyOut]
    alerts: list[MonitoringAlertOut]
    observations: list[BehaviorObservationOut]
    intervention_ids: list[str]


class BehaviorInterventionRecordOut(BaseModel):
    id: str
    status: BehaviorInterventionStatus
    intervention_code: str
    name: str
    trigger_biases: list[str]
    scenario_code: str
    personalized_message: str
    action_instruction: str
    cooling_period_hours: int | None
    starts_at: datetime
    eligible_at: datetime | None
    evidence: dict[str, Any]


class BehaviorInterventionsResponse(BaseModel):
    household_id: str
    interventions: list[BehaviorInterventionRecordOut]


class NextBestActionOut(BaseModel):
    action_code: str
    household_id: str
    priority: int = Field(ge=0)
    trigger_reason: str
    client_impact: str
    recommended_action: str
    do_not_sell_flag: bool
    required_specialist: str | None
    evidence: dict[str, Any]
    due_date: date | None


class NextBestActionsResponse(BaseModel):
    household_id: str
    outcome: Literal["ACTIONS_AVAILABLE", "NO_ACTION_REQUIRED"]
    actions: list[NextBestActionOut]
    boundary: str


class AdvisorActionCenterItem(BaseModel):
    trigger_id: str
    household_id: str
    household_code: str
    household_name: str
    trigger_type: str
    urgency: MonitoringSeverity
    reason: str
    required_role: str
    follow_up_due: date | None
    status: AdvisorTriggerStatus
    action_item_id: str | None
    action_code: str | None
    action_type: str | None
    title: str | None
    do_not_sell_flag: bool
    required_specialist: str | None
    evidence: dict[str, Any]


class AdvisorActionCenterResponse(BaseModel):
    items: list[AdvisorActionCenterItem]
    open_count: int
    overdue_count: int
    boundary: str

File: frontend/src/api/monitoring.ts
import { actorHeaders, demoActor } from "./actor";

export type MonitoringSeverity = "info" | "watch" | "high" | "critical";

export interface MonitoringAlert {
  id: string;
  household_id: string;
  monitoring_policy_id: string;
  financial_event_id: string | null;
  policy_type: string;
  trigger_reason: string;
  client_impact: string;
  severity: MonitoringSeverity;
  recommended_action: string;
  do_not_sell_flag: boolean;
  required_specialist: string | null;
  evidence_snapshot: Record<string, unknown>;
  status: "open" | "acknowledged" | "resolved" | "dismissed";
  detected_at: string;
  resolved_at: string | null;
}

export interface MonitoringAlertsResponse {
  household_id: string;
  rule_version: string;
  alerts: MonitoringAlert[];
}

export interface NextBestAction {
  action_code: string;
  household_id: string;
  priority: number;
  trigger_reason: string;
  client_impact: string;
  recommended_action: string;
  do_not_sell_flag: boolean;
  required_specialist: string | null;
  evidence: Record<string, unknown>;
  due_date: string | null;
}

export interface NextBestActionsResponse {
  household_id: string;
  outcome: "ACTIONS_AVAILABLE" | "NO_ACTION_REQUIRED";
  actions: NextBestAction[];
  boundary: string;
}

export interface AdvisorActionCenterItem {
  trigger_id: string;
  household_id: string;
  household_code: string;
  household_name: string;
  trigger_type: string;
  urgency: MonitoringSeverity;
  reason: string;
  required_role: string;
  follow_up_due: string | null;
  status: "open" | "acknowledged" | "resolved" | "dismissed";
  action_item_id: string | null;
  action_code: string | null;
  action_type: string | null;
  title: string | null;
  do_not_sell_flag: boolean;
  required_specialist: string | null;
  evidence: Record<string, unknown>;
}

export interface AdvisorActionCenterResponse {
  items: AdvisorActionCenterItem[];
  open_count: number;
  overdue_count: number;
  boundary: string;
}

export const monitoringFeatureEnabled =
  import.meta.env.VITE_ENABLE_V5_MONITORING === "true";

const apiBase = (import.meta.env.VITE_API_BASE_URL ?? "").replace(/\/$/, "");

async function request<T>(
  path: string,
  role: "client" | "advisor",
  signal?: AbortSignal,
): Promise<T> {
  const response = await fetch(apiBase + path, {
    signal,
    headers: {
      Accept: "application/json",
      ...actorHeaders(demoActor(role)),
    },
  });
  if (!response.ok) {
    const payload = await response.json().catch(() => null) as {
      error?: { message?: string };
    } | null;
    throw new Error(payload?.error?.message ?? `请求失败（${response.status}）`);
  }
  return await response.json() as T;
}

function householdPath(householdId: string): string {
  return `/api/v1/households/${encodeURIComponent(householdId)}`;
}

export function fetchMonitoringAlerts(
  householdId: string,
  signal?: AbortSignal,
): Promise<MonitoringAlertsResponse> {
  return request(`${householdPath(householdId)}/monitoring/alerts`, "client", signal);
}

export function fetchNextBestActions(
  householdId: string,
  signal?: AbortSignal,
): Promise<NextBestActionsResponse> {
  return request(`${householdPath(householdId)}/next-best-actions`, "client", signal);
}

export function fetchAdvisorActionCenter(
  signal?: AbortSignal,
): Promise<AdvisorActionCenterResponse> {
  return request("/api/v1/advisor/action-center", "advisor", signal);
}
